[FLINK-21239][Table SQL/API] Upgrade Calcite to 1.28.0

snuyanzin opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

*(For example: This pull request makes task deployment go through the blob server, rather than through RPC. That way we avoid re-transferring them on each deployment (during recovery).)*


## Brief change log

*(for example:)*
  - *The TaskInfo is stored in the blob store on job creation time as a persistent artifact*
  - *Deployments RPC transmits only the blob storage reference*
  - *TaskManagers retrieve the TaskInfo from the blob cache*


## Verifying this change

Please make sure both new and modified tests in this PR follows the conventions defined in our code quality guide: https://flink.apache.org/contributing/code-style-and-quality-common.html#testing

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (100MB)*
  - *Extended integration test for recovery after master (JobManager) failure*
  - *Added test that validates that TaskInfo is transferred only once across recoveries*
  - *Manually verified the change by running a 4 node cluster with 2 JobManagers and 4 TaskManagers, a stateful streaming program, and killing one JobManager and two TaskManagers during the execution, verifying that recovery happens correctly.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / no)
  - The serializers: (yes / no / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / no / don't know)
  - The S3 file system connector: (yes / no / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

snuyanzin commented on flink-table/flink-sql-client/src/test/resources/sql/module.q:
```diff
@@ -63,6 +63,7 @@ Was expecting one of:
     <BRACKET_QUOTED_IDENTIFIER> ...
     <QUOTED_IDENTIFIER> ...
     <BACK_QUOTED_IDENTIFIER> ...
+    <BIG_QUERY_BACK_QUOTED_IDENTIFIER> ...
```
As a result of https://issues.apache.org/jira/browse/CALCITE-4847

snuyanzin commented on flink-table/flink-table-planner/src/main/java/org/apache/calcite/sql2rel/RelDecorrelator.java:
```diff
@@ -128,11 +129,9 @@
- *   <li>Was changed within FLINK-29280, FLINK-28682: Line 224 ~ 234
- *   <li>Should be removed after fix of FLINK-29540: Line 296 ~ 302
- *   <li>Should be removed after fix of FLINK-29540: Line 314 ~ 320
- *   <li>Was changed within FLINK-21592: Line 1954 ~ 1962, Should be removed after update to Calcite
- *       1.28.0 as it is fixed at CALCITE-4773
```
This Flink modification is removed

snuyanzin commented:
@flinkbot run azure

twalthr commented on flink-table/flink-table-planner/src/main/java/org/apache/calcite/tools/RelBuilder.java:
```diff
@@ -0,0 +1,5161 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.tools;
```
We offer a `FlinkRelBuilder` already. Can we override the affected methods instead of copying the entire `RelBuilder` class? It seems only the

```
public RelBuilder join(JoinRelType joinType, RexNode condition,
      Set<CorrelationId> variablesSet) {
```

Is affected?

snuyanzin replied:
yes, only this method is affected.
The problem with `FlinkRelBuilder` is that `org.apache.calcite.tools.RelBuilder#join(org.apache.calcite.rel.core.JoinRelType, org.apache.calcite.rex.RexNode, java.util.Set<org.apache.calcite.rel.core.CorrelationId>)` operates with private fields `stack`, `simplifier`, `config`, `struct` and private static classes `org.apache.calcite.tools.RelBuilder.Frame`, `org.apache.calcite.tools.RelBuilder.Shifter` which are not reachable from `FlinkRelBuilder`... 
At the same time probably it could be a point to improve it on Calcite side to make it extendable in future

snuyanzin replied:
I added a follow up task for that https://issues.apache.org/jira/browse/CALCITE-5381
Once it is fixed this class could be removed
By the way: I've checked it till 1.32.0 and this class does not require modification from Flink to make it working

twalthr reviewed (changes requested):
Great work @snuyanzin. Also nice that no tests had to be touched during the upgrade. My biggest concern is regarding pom.xml files so far. Maybe we can further simplify, add exclusions, and comment what is really necessary for Flink.

snuyanzin commented on flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/AbstractCodeGeneratorCastRule.java:
```diff
@@ -204,9 +204,10 @@ public String getSessionTimeZoneTerm() {
 
         @Override
-        public String declareVariable(String type, String variablePrefix) {
+        public String declareVariable(String type, String variablePrefix, String defaultValue) {
```
yes that's possible.
However it's not only change in pom, after https://issues.apache.org/jira/browse/CALCITE-3745 it uses new api for compiler factory initialization. To use janino 3.0.11 need to rollback this (includes 
`EnumerableInterpretable.java`,`JaninoRexCompiler.java`, `JaninoRelMetadataProvider.java`,
`JaninoCompiler.java`) 
the commit will look like https://github.com/snuyanzin/flink/commit/0fd02d8ab9ac589b8d1bab9b92740bba193a2bed
and other 3 commits from this PR workarounding janino's issues could be dropped then

snuyanzin replied:
I added these mentioned changes making Flink working with Calcite 1.28.0 and Janino 3.0.11

snuyanzin commented on flink-core/src/main/java/org/apache/flink/core/classloading/ComponentClassLoader.java:
```diff
@@ -189,6 +190,14 @@ public URL getResource(final String name) {
 
+    @Override
+    public InputStream getResourceAsStream(String name) {
+        if (isOwnerFirstClass(name)) {
```
yes, probably you are right, done.
Initially I thought that there is no other usages of that methods,  however it worth to keep similar structure

snuyanzin commented on flink-table/flink-sql-parser-hive/pom.xml:
```diff
@@ -63,11 +63,51 @@ under the License.
 			<groupId>org.apache.calcite</groupId>
 			<artifactId>calcite-core</artifactId>
 			<version>${calcite.version}</version>
 			<exclusions>
```
yes, sure
`mvn dependency:tree` added

twalthr replied:
Shouldn't `org.codehaus.janino:commons-compiler:jar:3.1.8` and `org.codehaus.janino:janino:jar:3.1.8` not be part of the `dependency:tree` output?

snuyanzin commented on flink-table/flink-table-planner/pom.xml:
```diff
@@ -49,6 +49,18 @@ under the License.
 				</exclusion>
 			</exclusions>
 		</dependency>
+		<dependency>
```
Ah, yes, forgot to mention it.
This is annotation processor which should replace custom Calcite immutable beans (started at https://issues.apache.org/jira/browse/CALCITE-4787).

This is only compile time dependency (no runtime).
It is required for some classes copied from Calcite like `RelDecorrelator`, `SqlToRelConverter`

snuyanzin replied:
so I guess no relocation and mentioning in NOTICE is required.
May be a clarification comment in pom

snuyanzin replied:
also piece of `mvn dependency:tree` with `org.immutables` looks like
```
[INFO] --- maven-dependency-plugin:3.2.0:tree (default-cli) @ flink-table-planner_2.12 ---
[INFO] org.apache.flink:flink-table-planner_2.12:jar:1.17-SNAPSHOT
...
[INFO] +- org.immutables:value:jar:2.8.8:compile
[INFO] +- org.immutables:value-annotations:jar:2.8.8:compile
...
```

snuyanzin commented:
@twalthr thanks for your feedback
I adapted PR to use Janino 3.0.11 (however it requires to copy/paste and change 4 Calcite classes which could be removed once Janino is upgraded) and addressed your feedback.
Also I noticed that there is an existing JIRA issue to upgrade Janino https://issues.apache.org/jira/browse/FLINK-27995

snuyanzin commented:
@flinkbot run azure

snuyanzin commented:
@flinkbot run azure

snuyanzin commented:
Current failure is related to https://issues.apache.org/jira/browse/FLINK-24119
There are some fixes however this branch does not include them since it is not rebased...

twalthr reviewed (approved):
Thanks for the update @snuyanzin. Great that the PR now basically only consists of copied Calcite classes and pom.xml changes. I will now check out the PR a final time and go over the dependency trees to be on the safe side. I will merge this afterwards (likely today).

twalthr commented:
@flinkbot run azure
